app/helpers/random_assign_helper.rb

```diff
@@ -5,186 +5,234 @@ class UnableToRandomlyAssignGroupException < Exception
+      reviewee_id = @shuffled_reviewees[swap_index]
+      if can_assign?(reviewer_id, reviewee_id)
+        do_forward_assign_swap_and_mark(reviewer_id, reviewee_id, shuffle_index, swap_index)
+        return true
+      end
     end
 
-    # Since they're not assigned, the returned boolean is if they don't share students.
-    reviewer.does_not_share_any_students?(reviewee)
+    false
   end
 
-  def swap_shuffled_indices(first_index, second_index)
-    @shuffled_reviewees[first_index], @shuffled_reviewees[second_index] =
-        @shuffled_reviewees[second_index], @shuffled_reviewees[first_index]
-  end
+  def can_assign_to_prev_index?(reviewer_id, prev_index, shuffle_index)
+    current_reviewee_id = @shuffled_reviewees[shuffle_index]
+    prev_reviewer_id = @assigned_reviewer_to_shuffled_reviewee[prev_index]
+    prev_reviewee_id = @shuffled_reviewees[prev_index]
 
-  def add_peer_review_to_db_and_remember_assignment(reviewer, reviewee, shuffle_index)
-    peer_review = PeerReview.create_peer_review_between(reviewer, reviewee)
-    @reviewers_assigned_to[reviewer.id].add(reviewee.id)
-    @shuffled_reviewees_pr[shuffle_index] = peer_review
+    # We can only backward swap assign if the prev/current reviewee are different,
+    # it's not the same group, and if they can be assigned properly after swapping.
+    current_reviewee_id != prev_reviewee_id && reviewer_id != prev_reviewer_id &&
+        can_assign?(reviewer_id, prev_reviewee_id) && can_assign?(prev_reviewer_id, current_reviewee_id)
   end
 
-  # Goes forward from shuffle_index + 1 constantly ahead and either swaps on
-  # finding an eligible reviewee from wrap-around (uses modulus) or throws
-  # an UnableToRandomlyAssignGroupException.
-  def find_and_swap_with_group_or_throw(reviewer, shuffle_index)
-    next_index = 0
-    reviewee_at_shuffle_index = @shuffled_reviewees[shuffle_index]
-    while next_index < shuffle_index
-      reviewee_at_previous_index = @shuffled_reviewees[next_index]
-      reviewer_assigned_to_previous_reviewee = @shuffled_reviewees_pr[next_index].reviewer
+  def swap_previous_with_current_and_assign(reviewer_id, prev_index, shuffle_index)
+    prev_reviewer_id = @assigned_reviewer_to_shuffled_reviewee[prev_index]
 
-      unless eligible_to_be_assigned(reviewer, reviewee_at_previous_index) and
-             eligible_to_be_assigned(reviewer_assigned_to_previous_reviewee, reviewee_at_shuffle_index)
-        next_index = (next_index + 1) % @shuffled_reviewees.size
-        next
-      end
+    unmark_reviewer_assignment(prev_index)
+    swap_indices(@shuffled_reviewees, prev_index, shuffle_index)
+    mark_reviewer_assigned_to(prev_reviewer_id, @shuffled_reviewees[prev_index], prev_index)
+    mark_reviewer_assigned_to(reviewer_id, @shuffled_reviewees[shuffle_index], shuffle_index)
+  end
 
-      perform_group_exchange(reviewer, shuffle_index, next_index)
-      return
+  def attempt_assign_backwards(reviewer_id, shuffle_index)
```

Comment:
That was me trying to skip out on using the modulus and seeing if there was a performance boost (I don't think there was or it was negligible, this is probably what I get for spending a large chunk of my free time reading assembly optimizations).
This would be converted back, but was just done as a test to see if it helped at all.